shared/src/ledger/ibc/connection.rs

```diff
@@ -0,0 +1,423 @@
+//! IBC validity predicate for connection module
+
+use std::str::FromStr;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use ibc::ics02_client::client_consensus::{AnyConsensusState, ConsensusState};
+use ibc::ics02_client::client_state::AnyClientState;
+use ibc::ics02_client::context::ClientReader;
+use ibc::ics02_client::height::Height;
+use ibc::ics03_connection::connection::{ConnectionEnd, Counterparty, State};
+use ibc::ics03_connection::context::ConnectionReader;
+use ibc::ics03_connection::handler::verify::verify_proofs;
+use ibc::ics07_tendermint::consensus_state::ConsensusState as TendermintConsensusState;
+use ibc::ics23_commitment::commitment::CommitmentPrefix;
+use ibc::ics24_host::identifier::{ClientId, ConnectionId};
+use ibc::ics24_host::Path;
+use ibc::proofs::Proofs;
+use tendermint_proto::Protobuf;
+
+use super::{Error, Ibc, Result, StateChange};
+use crate::ledger::storage::{self, StorageHasher};
+use crate::types::address::{Address, InternalAddress};
+use crate::types::ibc::{
+    ConnectionOpenAckData, ConnectionOpenConfirmData, ConnectionOpenTryData,
+    Error as IbcDataError,
+};
+use crate::types::storage::{BlockHeight, Epoch, Key, KeySeg};
+
+pub(super) const COUNTER_PATH: &str = "connections/counter";
+
+impl<'a, DB, H> Ibc<'a, DB, H>
+where
+    DB: 'static + storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: 'static + StorageHasher,
+{
+    pub(super) fn validate_connection(
+        &self,
+        key: &Key,
+        tx_data: &[u8],
+    ) -> Result<bool> {
+        if Self::is_counter_key(key) {
+            // the counter should be increased
+            return Ok(
+                self.connection_counter_pre() < self.connection_counter()
```

Comment:
should we check that it is exactly `+1` or is it enough to check that it is greater?

Reply:
I guess it can be greater than `+1` when there is more than 1 new connection in the same tx, right?

Reply:
That's right. Multiple connections can be created in the same tx.